src/frontend/src/optimizer/rule/table_function_to_project_set_rule.rs

```diff
@@ -54,39 +54,51 @@ impl Rule for TableFunctionToProjectSetRule {
             logical_table_function.base.ctx.clone(),
         );
         let logical_project_set = LogicalProjectSet::create(logical_values, vec![table_function]);
-        // We need a project to align schema type because `LogicalProjectSet` has a hidden column
-        // `projected_row_id` and table function could return multiple columns, while project set
-        // return only one column with struct type.
+        // We need a project to align schema type because
+        // 1. `LogicalProjectSet` has a hidden column `projected_row_id` (0-th col)
+        // 2. When the function returns a struct type, TableFunction will return flatten it into multiple columns, while ProjectSet still returns a single column.
         let table_function_col_idx = 1;
-        if let DataType::Struct(st) = table_function_return_type.clone() {
+        let logical_project = if let DataType::Struct(st) = table_function_return_type.clone() {
             let exprs = st
                 .types()
                 .enumerate()
                 .map(|(i, data_type)| {
                     let field_access = FunctionCall::new_unchecked(
                         ExprType::Field,
                         vec![
-                            ExprImpl::InputRef(
-                                InputRef::new(
-                                    table_function_col_idx,
-                                    table_function_return_type.clone(),
-                                )
-                                .into(),
-                            ),
+                            InputRef::new(
+                                table_function_col_idx,
+                                table_function_return_type.clone(),
+                            )
+                            .into(),
                             ExprImpl::literal_int(i as i32),
                         ],
                         data_type.clone(),
                     );
                     ExprImpl::FunctionCall(field_access.into())
                 })
                 .collect_vec();
-            let logical_project = LogicalProject::new(logical_project_set, exprs);
-            Some(logical_project.into())
+            LogicalProject::new(logical_project_set, exprs)
         } else {
-            let logical_project = LogicalProject::with_out_col_idx(
+            LogicalProject::with_out_col_idx(
                 logical_project_set,
                 std::iter::once(table_function_col_idx),
-            );
+            )
+        };
+
+        if logical_table_function.with_ordinality {
+            let projected_row_id = InputRef::new(0, DataType::Int64).into();
+            let ordinality = FunctionCall::new(
+                ExprType::Add,
+                vec![projected_row_id, ExprImpl::literal_bigint(1)],
```

Comment:
Since `projected_row_id ` starts from 0, we need to add one to align with `with ordinality `

Reply:
Indeed